[FAB-17540] Fix race TestUpdateRootsFromConfigBlock

stephyee opened the pull request:
#### Type of change

- Bug fix

#### Description

Lock read/write access to tlsconfig to prevent data race

#### Related issues

[FAB-17540](https://jira.hyperledger.org/browse/FAB-17540)

yacovm commented:
I am confused.

The stack trace in the JIRA shows a race in the production code, where Fabric does:
```
WARNING: DATA RACE
Write at 0x00c001242690 by goroutine 216:
 github.com/hyperledger/fabric/core/comm.(*GRPCServer).SetClientRootCAs()
      /home/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/core/comm/server.go:293 +0x315
  github.com/hyperledger/fabric/core/peer.(*Peer).updateTrustedRoots()
      /home/sykesm/workspace/fabric/src/github.com/hyperledger/fabric/core/peer/peer.go:113 +0x5c4
```
And the TLS handshake does:

```
Previous read at 0x00c001242690 by goroutine 128:
  crypto/tls.(*Conn).processCertsFromClient()
      /usr/local/go/src/crypto/tls/handshake_server.go:730 +0x405
  crypto/tls.(*serverHandshakeState).doFullHandshake()
      /usr/local/go/src/crypto/tls/handshake_server.go:514 +0xa23
  crypto/tls.(*serverHandshakeState).handshake()
      /usr/local/go/src/crypto/tls/handshake_server.go:105 +0x68c
  crypto/tls.(*Conn).serverHandshake()
      /usr/local/go/src/crypto/tls/handshake_server.go:60 +0x1cd
```


But you are changing only the test code.

Don't we potentially have a problem in the production code? 

The way I see it, the Fabric code changes: **gServer.tlsConfig.ClientCAs** and the TLS handshake reads these TLS root CAs. 

Perhaps we should consider implementing a [TransportCredentials](https://godoc.org/google.golang.org/grpc/credentials#TransportCredentials) which will atomically load an immutable TLSConfig ?

cendhu commented:
@yacovm -- the data race expert :-)

stephyee commented:
@yacovm Is it preferred to add a mutex to the serverConfig? I used the [clone()](https://godoc.org/crypto/tls#Config.Clone) which uses a mutex.

yacovm reviewed (changes requested):
> @yacovm Is it preferred to add a mutex to the serverConfig? I used the [clone()](https://godoc.org/crypto/tls#Config.Clone) which uses a mutex.

```
func (c *Config) Clone() *Config {
	// Running serverInit ensures that it's safe to read
	// SessionTicketsDisabled.
	c.serverInitOnce.Do(func() { c.serverInit(nil) })
	var sessionTicketKeys []ticketKey
	c.mutex.RLock()
	sessionTicketKeys = c.sessionTicketKeys
	c.mutex.RUnlock()
```
It uses a mutex, but: 

1. It only synchronizes around the session ticket keys which is irrelevant to the client root CAs.
2. The **SetClientRootCAs** method which writes to the memory, doesn't lock the same mutex, so there is no [happens-before](https://go101.org/article/memory-model.html) relation enforced by the runtime.

It is true that the documentation says:
```
 It is safe to clone a Config that is being used concurrently by a TLS client or server. 
```
However this is misleading in our case, because the documentation didn't take into account that the Fabric code, changes the config as well. 

You need to make all writes and reads to the client roots to be thread safe and memory coherent.

There are 2 paths for you to take:

1. Wrap every access that writes to the tlsConfig field of the GRPCServer struct with a write lock, and wrap every read access to that struct with a read lock. This is less code but is more error prone in the long run.
2. Remove the tlsConfig field, and replace it with something that enforces the user to access it in a memory coherent manner. For instance, you can wrap the **tls.Config** with a struct with a method that returns a copy of the inner **tls.Config** (the copying is taken under a read lock) and a method that mutates the **tls.Config** under a write lock. An example can be seen in the [PredicateDialer](https://github.com/hyperledger/fabric/blob/master/orderer/common/cluster/util.go#L113-L123) of the cluster infrastructure of the orderer.

stephyee commented:
@yacovm thanks I'll consider both options you presented

yacovm commented on internal/pkg/comm/creds.go:
```diff
@@ -56,6 +58,9 @@ func (sc *serverCreds) ClientHandshake(context.Context,
 
 // ServerHandshake does the authentication handshake for servers.
 func (sc *serverCreds) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	sc.lock.Lock()
```
TLS handshakes involve network I/O so we don't want to hold locks for a long time. Even if there is nothing competing this lock, it is usually considered bad practice to hold locks while doing network I/O.

I suggest that the TLS config will be cloned under a read lock and nothing more.

stephyee committed:
[FAB-17540] Fix for race read/write tlsconfig

- Encapsulate tls.Config in a struct with a lock to prevent data races
when accesing tls.Config
- Fix TestCreds assertion on creds.Clone() was invalid because of
tls.Config's mutex

Signed-off-by: Danny Cao <dcao@us.ibm.com>
Signed-off-by: Tiffany Harris <tiffany.harris@ibm.com>

yacovm commented on internal/pkg/comm/creds.go:
```diff
@@ -28,26 +29,50 @@ var (
 
+type TLSConfig struct {
+	config *tls.Config
+	lock   *sync.Mutex
```
no need to have a pointer mutex. a non pointer is just fine, and saves you initialization.

stephyee replied:
I'm not sure about this, w/o the pointer this triggers `copylocks` check failure with govet.

yacovm replied:
But with your latest version of the code, you don't reference the lock externally anymore.

stephyee replied:
I'm not sure what you mean by this statement, the `AddRootCA()` and `SetClientCAs()` still reference the TLConfig lock.

yacovm replied:
Right, but these are all methods of the struct, so they reference it internally. 
By externally i mean that the lock is referenced by some other function which is not an instance method of the struct TLSConfig

yacovm commented on internal/pkg/comm/server.go:
```diff
@@ -73,11 +73,15 @@ func NewGRPCServerFromListener(listener net.Listener, serverConfig ServerConfig)
 	if secureConfig.UseTLS {
 		//both key and cert are required
 		if secureConfig.Key != nil && secureConfig.Certificate != nil {
+			grpcServer.tls = &TLSConfig{
```
you have a constructor method, yet you create the instance without it.

Let's be consistent.

yacovm commented on internal/pkg/comm/server.go:
```diff
@@ -245,6 +252,8 @@ func (gServer *GRPCServer) AppendClientRootCAs(clientRoots [][]byte) error {
 
 // internal function to add a PEM-encoded clientRootCA
 func (gServer *GRPCServer) appendClientRootCA(clientRoot []byte) error {
+	gServer.tls.lock.Lock()
```
this works, but is very brittle as no one promises you that the TLS config object is not going to be modified further in the code execution without grabbing the lock beforehand.

yacovm commented on internal/pkg/comm/creds.go:
```diff
@@ -28,26 +29,50 @@ var (
+	}
+}
+
+func (sc *TLSConfig) Config() *tls.Config {
```
I suggest that the TLSConfig struct would govern the lifecycle of its inner fields and not delegate it to the user of the struct:

1. Let's return a copy upon Config(). Note, that a shallow copy is trivially obtained via changing the return type to be a non pointer type.
2. Let's add a mutator method that just updates whatever we want, under the lock.

This removes the need to external locking completely and simplifies the usage of this struct.

yacovm reviewed (changes requested):
The comment above + I think it's nice to add a unit test.

yacovm commented on internal/pkg/comm/creds.go:
```diff
@@ -28,26 +30,64 @@ var (
 
+type TLSConfig struct {
+	config *tls.Config
+	lock   *sync.Mutex
```
Let's use a RWMutex so it will be non blocking for TLS hanshakes.

yacovm commented on internal/pkg/comm/creds.go:
```diff
@@ -28,26 +30,64 @@ var (
+}
+
+func (t *TLSConfig) Config() tls.Config {
+	t.lock.Lock()
```
Let's use a read lock here instead so you can do several TLS handshakes in parallel.

yacovm commented on internal/pkg/comm/creds.go:
```diff
@@ -28,26 +30,64 @@ var (
+	return tls.Config{}
+}
+
+func (t *TLSConfig) AddCert(cert *x509.Certificate) {
```
AddClientRootCA gives the method more meaning that AddCert

yacovm commented on internal/pkg/comm/server.go:
```diff
@@ -234,6 +237,7 @@ func (gServer *GRPCServer) Stop() {
 func (gServer *GRPCServer) AppendClientRootCAs(clientRoots [][]byte) error {
```
This method (**AppendClientRootCAs**) is actually not used in production code. 
Let's get rid of it shall we?

yacovm commented on internal/pkg/comm/creds.go:
```diff
@@ -28,26 +30,64 @@ var (
 	// NOTE: unlike the default grpc/credentials implementation, we do not
 	// clone the tls.Config which allows us to update it dynamically
-	serverConfig.NextProtos = alpnProtoStr
+	serverConfig.lock.Lock()
```
i might be missing something, but - why do we need to lock here?

yacovm commented on internal/pkg/comm/creds.go:
```diff
@@ -28,26 +30,64 @@ var (
+	lock   *sync.Mutex
+}
+
+func NewTLSConfig(config *tls.Config, lock *sync.Mutex) *TLSConfig {
```
I think that now you don't need to inject the lock anymore, because I don't think you need to lock in **NewServerTransportCredentials**.

yacovm reviewed (changes requested):
Some more comments

yacovm commented on internal/pkg/comm/creds.go:
```diff
@@ -28,26 +30,61 @@ var (
+	return tls.Config{}
+}
+
+func (t *TLSConfig) AddRootCA(cert *x509.Certificate) {
```
**AddClientRootCA** ? :-)

yacovm reviewed (approved):
Thanks for the patience and compliance to the code review comments.

yacovm commented:
I think this should be backported since it effects production behavior of TLS handshakes.
